appyters/GSEA_Appyter/GSEA_Appyter.ipynb

```diff
@@ -0,0 +1,728 @@
+    "    if (library_tab == 'Upload' and library_filename.endswith('.gmt') \n",
+    "    == False):\n",
+    "        raise Exception('Please upload a GMT file (ends in .gmt).')\n",
+    "\n",
+    "#performs GSEA\n",
+    "def gsea():\n",
+    "    if data_tab == 'Option 1':\n",
+    "        rnk = pd.read_csv(ranked_filename, header=None, sep=\"\\t\")\n",
+    "        results = gp.prerank(rnk=rnk, gene_sets=library_filename, max_size=500)\n",
+    "    else:\n",
+    "        results = gp.gsea(data=expression_filename, gene_sets=library_filename,\n",
+    "        cls=phenotype_filename, max_size=500)\n",
+    "    return results"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "beb3a14b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#loads a preranked list and converts it to a dataframe for Option 1\n",
+    "#creates a ranked list with chosen ranking method and converts to dataframe for Option 2\n",
+    "def loadRanked():\n",
+    "    if data_tab == 'Option 1':\n",
+    "        values = pd.read_csv(ranked_filename, header=None, sep=\"\\t\")\n",
+    "        if len(values.index) < 5000:\n",
+    "            raise Exception('Current ranked gene list has less than 5000 genes. We expect all human coding genes to be in the list, which is around 20000.')\n",
+    "        values.sort_values(by=values.columns[1], ascending=False, inplace=True)\n",
+    "        values.reset_index(drop=True, inplace=True)\n",
+    "        values.columns = ['Gene', 'Rank']\n",
+    "    elif data_tab == 'Option 2':\n",
+    "        pos, neg, classes = gp.parser.gsea_cls_parser(phenotype_filename)\n",
+    "        obj = GSEA(data=expression_filename, gene_sets=library_filename,\n",
+    "        classes=classes)\n",
+    "        exp, classesDict = obj.load_data(classes)\n",
+    "        values = gp.algorithm.ranking_metric(df=exp, pos=pos, neg=neg, \n",
+    "        method=ranking_method, classes=classesDict, ascending=False)\n",
+    "        values = pd.DataFrame({'Gene':values.index, 'Rank':values.values})\n",
+    "    return values"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "bea4fe76",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#loads libraries\n",
+    "def downloadLibrary(name):\n",
+    "    with open(f\"{name}\", \"w\") as fw:\n",
+    "        with urllib.request.urlopen(f'https://maayanlab.cloud/Enrichr/geneSetLibrary?mode=text&libraryName={name}') as f:\n",
+    "            for line in f.readlines():\n",
+    "                fw.write(line.decode('utf-8'))\n",
+    "                fw.flush()\n",
+    "def loadLibrary(library_filename):\n",
+    "    if library_tab == 'Select an Enrichr Library':\n",
+    "        downloadLibrary(library_filename)\n",
+    "    library_data = dict()\n",
+    "    with open(library_filename, 'r') as f:\n",
+    "        lines = f.readlines()\n",
+    "        for line in lines:\n",
```

Comment:
`for line in f:` eliminate `lines = f.readlines()` file objects in python are iterable